### PR TITLE
Also kill python process when terminating electron in Windows

### LINKED ISSUE #709: Windows client process fails to terminate on exit
## Problem Definition
[rotkehlchen.log](https://github.com/rotki/rotki/files/4175091/rotkehlchen.log)

Rotki Error

Upon exiting the app either via File > Exit or closing out the window, the rotkehlchen-1.1.1-windows.exe process located @ \AppData\Local\Temp\1XQmRojWPPWrpMdb5u2XbYRpZNj\resources\rotkehlchen_py_dist remains running.

On next startup, the app pops up the attached error after ~5 seconds. Terminating the rotkehlchen-1.1.1-windows.exe process resolves the issue.

Tested on Windows 10 Pro x64 Build 1909

Provide a description of what is the current problem and why you are raising this issue.
If it's a bug please describe what was the unexpected thing that occured and what was the
expected behaviour.

Rotkehlchen also logs debug information to the `rotkehlchen.log` file. Please attach it to the
issue as it may help us find the source of the issue faster.

### System Description

Here add a detailed description of your system, e.g. output of the following script:

```
uname -a
command -v geth && geth version
command -v parity && parity --version
command -v rotkehlchen && rotkehlchen version
[ -d .git ] && git rev-parse HEAD
```

### PR DESCRIPTION
**Pull request by LefterisJP**
Fix #709

@kelsos Before merging this I need you or someone else to confirm this "should work". I tried v1.1.1 in my Windows 7 desktop and I could not see the problem described in #709.

Can you try and see if this is reproducible? But looking at the code the part of killing the pyproc client was indeed missing in Windows and this PR should add it and in theory fix the problem, right?

Only setback is I have not reproduced it.

### CODE REVIEW

**Comment by kelsos** [comment]
Do you have a build I could test direct on windows 10?

**Comment by LefterisJP** [comment]
Test v1.1.1 and see if you can reproduce. Then I can give you a build with the PR's code.

**Comment by kelsos** [comment]
I did confirm already that this also happens to me on windows 10.

**Comment by LefterisJP** [comment]
Interesting. Does not happen to me for Windows 7. Or perhaps I did a mistake when testing, if we judge by the missing code when compared to v1.0.7.

I will ping you with the binary soon.

**Comment by LefterisJP** [comment]
@kelsos here: https://we.tl/t-ZxMQsUiGSw

**Comment by dsupport** [comment]
Issue still persists in v.1.1.12
As can be seen, there are two identical processes running. Upon program exit, only one is terminated.
FWIW the process with PID 12972 is the one that remains.

Rotki Error 2

**Comment by dsupport** [comment]
In v1.0.7 there are a total of four processes running, and upon exit all get terminated successfully.
Are you intending to open more than one of the same process? I believe now that in v1.1.1 you only have one rotki-win32.exe open it closes its respective pyproc client but not the duplicate.


Rotki Error 3

**Comment by LefterisJP** [comment]
@dsupport What you see is the python backend of Rotki. And due to the packager we use (pyinstaller) there is a quirk in Windows that creates two processes: https://github.com/pyinstaller/pyinstaller/issues/2483

I have no idea why in v1.0.7 you see 4 processes.

@kelsos So since @dsupport already checked it, what I did here is not sufficient. Can you also check and also have a look at the code? You may see something I don't. What's different in v1.0.7 and v1.1.0 handling of the python backend in the electron side?

**Comment by kelsos** [comment]
I am pretty sure that in all the rebasing we missed some commits because I remember spending time to fix that issue.

Basically we need to check were we call app quit. In some case it is actually called before the taskkill executes resulting in what we see

**Comment by kelsos** [comment]
Can you try removing the `app.exit()` that is after awaiting the py-handler termination on the `will-close` event?

**Comment by LefterisJP** [comment]
> I don't think removing the else is needed. We just need to ensure that app.exit() is not called before the taskkill command executes.

But then who kills the python backend?

**Comment by kelsos** [comment]
So what I think is happening is that this exit call is called early:
https://github.com/rotki/rotki/blob/bc9ffb92bcab502ab4a795003bea17950604219d/electron-app/src/background.ts#L108

Which means that the electron process is terminated before being able to spawn the taskkill processes here: https://github.com/rotki/rotki/blob/bc9ffb92bcab502ab4a795003bea17950604219d/electron-app/src/py-handler.ts#L250

If we remove that `app.exit()`, since taskkill execution should call `app.exit()` either on success or failure adding an `app.exit()` to the else clause would also be required.

https://github.com/rotki/rotki/blob/bc9ffb92bcab502ab4a795003bea17950604219d/electron-app/src/py-handler.ts#L130

So the idea is that when the user closes the application, we keep the electron process alive long enough for it to execute the taskkill.

Does this help?

**Comment by LefterisJP** [comment]
So @kelsos, check the diff. Instead of removing the `app.exit();` and adding it in the else, I conditionally removed it if it's windows. Should have the same effect as what you intended, correct?

And here is a binary I just cooked up: https://we.tl/t-XBiQabZPuu

**Comment by kelsos** [comment]
Other than getting an unhandle javascript exception popup the first time I executed it, it seems to work properly now
```
2020-02-09T22:26:07.095Z: Detected the following running python sub-processes: 15636, 7760
2020-02-09T22:26:07.100Z: Preparing to call "taskill /f /t /PID 15636 /PID 7760" on the python sub-processes
2020-02-09T22:26:07.205Z: The Python sub-process exited with signal: null (Code: 1)
2020-02-09T22:26:07.210Z: Call to taskkill exited
```

**Comment by dsupport** [comment]
Working on my end as well.
Here's what I see now process wise just for reference sake:

Rotki 4

**Comment by LefterisJP** [comment]
> Other than getting an unhandle javascript exception popup (only the first time I executed it, could not get it in the 5-6 tried afterward), it seems to work properly now

I also got a "network error" in the frontend the first time I executed it in Windows 7. No idea what that was. Did you by any chance get a stack trace to see what happened @kelsos ?

So I guess this does solve the problem seen and should be included as solution in the upcoming v1.2.0 as is. We will ofcourse test the final binary further, but need to know if the code change is good   (approve or not). You can take your time to approve @kelsos. v1.2.0 will not happen until most of the planned features are ready.

Thank you for confirming also @dsupport! So cool to have such quick feedback.

**Comment by kelsos** [comment]
@LefterisJP I also got the network error in one of the calls. No, unfortunately, there was no stack trace. This was a dialog by Electron that had no useful information. I will give it a try today running on the development vm to see if I can replicate it somehow.

File: electron-app/src/py-handler.ts [hunk 1]
```diff
@@ -202,6 +225,13 @@ export default class PyHandler {
       args.push('--api-cors', this._corsURL);
     }
     args.push('--logfile', path.join(this.logsPath, 'rotkehlchen.log'));
+    args = ['--api-port', port.toString()].concat(args);
+    this.logToFile(
+      'Starting packaged python subprocess: ' +
+        executable +
```
Comment (kelsos) [inline electron-app/src/py-handler.ts]: Why not use templates instead?

File: electron-app/src/py-handler.ts [hunk 2]
```diff
@@ -178,22 +183,40 @@ export default class PyHandler {
...
-      this.logAndQuit('No files found in the dist directory');
+      this.logAndQuit('ERROR: No files found in the dist directory');
     } else if (files.length !== 1) {
-      this.logAndQuit('Found more than one file in the dist directory');
+      this.logAndQuit('ERROR: Found more than one file in the dist directory');
     }
     let executable = files[0];
     if (!executable.startsWith('rotkehlchen-')) {
       this.logAndQuit(
-        `Unexpected executable name "${executable}" in the dist directory`
```
Comment (kelsos) [inline electron-app/src/py-handler.ts]: This is a template if you wonder.

File: electron-app/src/py-handler.ts [hunk 3]
```diff
@@ -178,22 +183,40 @@ export default class PyHandler {
...
     }
+    // in some systems the virtualenv's python is not detected from inside electron and the
+    // system python is used. Electron/node seemed to add /usr/bin to the path before the
+    // virtualenv directory and as such system's python is used. Not sure why this happens only
+    // in some systems. Check again in the future if this happens in Lefteris laptop Archlinux.
+    // To mitigate this if a virtualenv is detected we add its bin directory to the start
+    // start of the path
+    if (process.env.VIRTUAL_ENV) {
+      this.logToFile('Setting path');
+      process.env.PATH = process.env.VIRTUAL_ENV + '/bin:' + process.env.PATH;
```
Comment (LefterisJP) [inline electron-app/src/py-handler.ts]: looking at it it dawned on me I should use `path.sep` here @kelsos

File: electron-app/src/py-handler.ts (hunk 1 above)
Comment (LefterisJP) [reply electron-app/src/py-handler.ts]: done

**Comment by LefterisJP** [comment]
@kelsos so I tested this in windows on my side and all is well. It also fixes #698. When you find time can you check if the clear intervals address the issue you mentioned in Windows 10? Here is a binary I just made in Windows: https://we.tl/t-w6ij7rut7i

@dsupport if you see this and also want to test maybe wait until v1.2.0 or at least try with an account that is not your main one. This version will ugprade your DB (as it contains labels and tags) and then you won't be able to use v1.1.1 anymore.

**Pull request merged**